### PR TITLE
Remove specialized timeValueField() methods in XContentBuilder

### PR DESCRIPTION
**Pull request by tlrx**
This PR renames some specialized method like 

`timeValueField(String rawFieldName, String readableFieldName, TimeValue timeValue)`

in favor of a bit more concise method:

`field(String name, TimeValue timeValue)`

The method takes care of writing the human readable field and also the time value field using a default `TimeUnit` configured at the `XContentBuilder` level. This ensures that human readable field and time value fields share the exact same prefix and allows to build objects using a specific time unit.

Note: this PR breaks some field names that were not fully coherent. If agreed, I'll document them as breaking change before merging.

### CODE REVIEW

**Comment by tlrx** [comment]
Actually I'd like to clean up all specialized method like bytesize and dates on the same way, as well as cleaning other methods to better use jackson 2.8.1 features and have a clean execution path...

File: core/src/main/java/org/elasticsearch/common/xcontent/XContentBuilder.java [hunk 1]
```diff
@@ -70,6 +69,8 @@ public static XContentBuilder builder(XContent xContent, Set<String> includes, S
 
     private boolean humanReadable = false;
 
+    private TimeUnit timeUnit = TimeUnit.MILLISECONDS;
```
Comment (nik9000) [inline core/src/main/java/org/elasticsearch/common/xcontent/XContentBuilder.java]: Allowing this to be customizable is nice but it is going to break parsing pretty hard if this doesn't stay MILLISECONDS.

Comment (nik9000) [inline core/src/main/java/org/elasticsearch/common/xcontent/XContentBuilder.java]: I think it might be worth dropping the `timeUnit` for now? I'm just scared of it.

File: core/src/main/java/org/elasticsearch/common/xcontent/XContentBuilder.java [hunk 2]
```diff
@@ -512,29 +525,73 @@ public XContentBuilder rawValue(BytesReference content) throws IOException {
...
+                    raw = value.days();
+                }
+                break;
+            default:
+                throw new IllegalArgumentException("Unsupported time unit");
+        }
+        return field(name + suffix).value(raw);
+    }
+
+    public XContentBuilder value(TimeValue value) throws IOException {
```
Comment (nik9000) [inline core/src/main/java/org/elasticsearch/common/xcontent/XContentBuilder.java]: I'm not sure this is ever the right thing to do though - it'll spit out only the user readable version.

Also I think `field(String, Object)` will need so work so it forwards to your new implementations.

I think `value(Object)` will need some work as well. I think maybe it is right to just disallow `TimeValue` there. Make it only ok to use `TimeValue` with `field` variants.

**Comment by nik9000** [comment]
> Actually I'd like to clean up all specialized method like bytesize and dates on the same way

That sounds like a good plan. I left some comments one the specifics but the plan of removing the specialized variants sounds good to me.

File: core/src/main/java/org/elasticsearch/common/xcontent/XContentBuilder.java (hunk 1 above)
Comment (tlrx) [inline core/src/main/java/org/elasticsearch/common/xcontent/XContentBuilder.java]: I see that differently. I like the idea to allow unit to be configurable and I think it helps us to be more coherent in field names as well as better identify the places where we expect a given time unit because of further parsing.

File: core/src/main/java/org/elasticsearch/common/xcontent/XContentBuilder.java (hunk 2 above)
Comment (tlrx) [inline core/src/main/java/org/elasticsearch/common/xcontent/XContentBuilder.java]: I agree - I pushed a bit quickly on this.

File: core/src/main/java/org/elasticsearch/common/xcontent/XContentBuilder.java (hunk 1 above)
Comment (nik9000) [inline core/src/main/java/org/elasticsearch/common/xcontent/XContentBuilder.java]: Yeah - it'd be super nice to be able to set the time unit on the URL and get it back - I'm quite happy at the idea. I'm just worried about the times when we persist stuff in XContent form and it lives for a while. We'd have to expand the places where we parse quite a bit. I think we could to it fairly easily for ObjectParser and friends but all the old places where we parse by hand are going to be a pain. They'll work now just because nothing ever changes this from millis.

That is why I propose moving it out of this - if you just did the names cleanup that'd be a good thing but supporting more units is going to be a bigger change.

File: core/src/main/java/org/elasticsearch/common/xcontent/XContentBuilder.java [hunk 3]
```diff
@@ -512,29 +525,73 @@ public XContentBuilder rawValue(BytesReference content) throws IOException {
...
+                if (value != null) {
+                    raw = value.hours();
+                }
+                break;
+            case DAYS:
+                suffix = "_in_days";
+                if (value != null) {
+                    raw = value.days();
+                }
+                break;
```
Comment (javanna) [inline core/src/main/java/org/elasticsearch/common/xcontent/XContentBuilder.java]: is this switch really needed? Is that because we don't always print out time in millis when non readable?

Comment (nik9000) [inline core/src/main/java/org/elasticsearch/common/xcontent/XContentBuilder.java]: See the comments above - @tlrx wants to allow users to specify that they want it in something other than millis. It is a great idea but I think it should be another PR, separate from and coming after removing the specialized methods.

Comment (javanna) [inline core/src/main/java/org/elasticsearch/common/xcontent/XContentBuilder.java]: I saw the comments, but I wonder where this is needed.

**Comment by rjernst** [comment]
@tlrx Do you want to bring this PR up to date with master?

**Comment by dakrone** [comment]
@tlrx ping again about this, are you still interested in pursuing this?
